lib/spack/spack/database.py

```diff
@@ -1257,6 +1265,7 @@ def _decrement_ref_count(self, spec):
 
         rec = self._data[key]
         rec.ref_count -= 1
+        self.modified = True
```

Comment:
Sometimes you put `self.modified = True` at the end (after all modifications), and (like in this case) it occurs in the middle. While it technically doesn't matter given the semantics of `_write` (which writes nothing in the case of an Exception), it would still be nice to have this marker in a consistent spot (maybe even a decorator?).

Reply:
In some cases there are early returns, so replacing `self.modified = True` with a context manager or a function doesn't make a difference. I guess the better solution is to recognize changes in install records w/o a boolean; e.g. copy on write or so. But I don't think the current implementation is that bug prone

Reply:
Actually I take that back, it's pretty bug prone. I forgot to set `self.modified = False` on `_read` too.

Reply:
This boolean is not The Way :tm: